### PR TITLE
Push initial implementation

### PR DESCRIPTION
**Pull request by avdgrinten**
This PR adds the initial version of cbuildrt.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+			.required(true)
+		)
+		.get_matches();
+
+	let cfg_f = File::open(matches.value_of("cbuild-json").unwrap())
+		.expect("unable to open cbuild.json");
+	let cfg : serde_json::Value = serde_json::from_reader(cfg_f)
+		.expect("failed to parse cbuild.json");
+
+	return Config{
```
Comment (64) [inline src/main.rs:36]: you could use [`serde_derive`](https://serde.rs/derive.html) instead of doing this manually

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+				dest: PathBuf::from(bm["destination"].as_str().unwrap()),
+				src: PathBuf::from(bm["source"].as_str().unwrap())
+			}})
+			.collect::<Vec<_>>()
+	};
+}
+
+// Concatenates lhs and rhs as-if the rhs was a relative path.
+fn concat_absolute<L: AsRef<Path>, R: AsRef<Path>> (lhs: L, rhs: R) -> PathBuf {
+	let mut buf = lhs.as_ref().to_path_buf();
```
Comment (64) [inline src/main.rs:56]: I think this function can be `lhs.concat(rhs.strip_prefix("/").unwrap())` or something

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+		buf.push(c);
+	}
+	return buf;
+}
+
+fn run_init(cfg: &Config) -> ! {
+	// Create regular files in /dev such that we can bind mount the real devices later.
+	let dev_overlays = vec!("tty", "null", "zero", "full", "random", "urandom");
+	for f in &dev_overlays {
+		let _ = File::create(concat_absolute(&cfg.rootfs, "/dev/").join(f))
```
Comment (64) [inline src/main.rs:68]: You don't need `let _ =` here

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+			.expect("failed to wait for children");
+		if let nix::sys::wait::WaitStatus::Exited(pid, code) = child_status {
+			if pid == child_pid {
+				if code != 0 {
+					println!("child returned non-zero exit code");
+				}
+				exit(code);
+			}
+		}
+	}
```
Comment (64) [inline src/main.rs:162]: This whole chunk can be a `match`

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	nix::sched::unshare(
+		nix::sched::CloneFlags::CLONE_NEWUSER
+		| nix::sched::CloneFlags::CLONE_NEWPID
+	).expect("failed to unshare()");
+
+	// Write the uid_map and gid_map files.
+	let mut setgroups_f = File::create("/proc/self/setgroups")
+		.expect("unable to open setgroups");
+	setgroups_f.write_all("deny".as_bytes())
+		.expect("unable to write setgroups file");
```
Comment (64) [inline src/main.rs:183]: You could use [`std::fs::write`](https://doc.rust-lang.org/std/fs/fn.write.html) here instead

**Comment by 64** [comment]
Seems good, I'd also give it a quick `cargo fmt` and `cargo clippy` (I think clippy just complains about use of `return x;` at the end of functions which is a bit unidiomatic)

**Comment by avdgrinten** [comment]
Fixed all comments.

**Comment by 64** [comment]
LGTM

**Pull request merged**